keps/sig-api-machinery/5366-graceful-leader-transition/README.md

```diff
@@ -0,0 +1,870 @@
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Many high-availability (HA) Kubernetes components, including kube-scheduler and
+kube-controller-manager, rely on the leader election library in client-go. The
+current implementation mandates that when a component loses its leader lock, it
+must shut down immediately. This is typically handled by calling
+`klog.FlushAndExit()` in the `OnStoppedLeading` callback.
+
+When the leadership is lost, the component shuts down and waits for the kubelet
+to detect that the component is unhealthy and restart it. This has several
+significant drawbacks:
+
+- High Overhead: Restarting a component process incurs unnecessary computational
+  overhead and increases latency during a leadership transition
+- No Graceful Shutdown: The immediate call to `os.Exit()` prevents any graaceful
+  shutdown or cleanup operations.
+
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Allow leader-elected components to transition to a follower state without restarting the process
+- Enable graceful handling of leader lock loss
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- This KEP allows work towards graceful shutdown of controllers, but the actual
+  mechanisms of how we do graceful shutdown is outside the scope of this KEP.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The main control loops for `kube-controller-manager`, `kube-sceduler`, and
+`cloud-controller-manager` will be updated to support graceful leader
+transitions. When a leader fails to renew its lease, instead of exiting, the
+component will rely on client-go's leader election mechanism to cancel the
+context, and stop its internal controllers. It will then immediately return to a
+follower state where it will attempt to reacquire the lease.
+
+This change will be guarded by a new feature gate, `GracefulLeaderTransition`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+Cloud provider A wants to balance controllers over multiple control planes. With
+graceful transitions, multiple locks can be used by a single KCM instance such
+that a subset of components run under each lock.
+
+### Risks and Mitigations
```

Comment:
Can we expand on this?  I'd like to make sure we're covering both:

- More detailed breakdown of the risks into specific failure modes (leaks that cause resource exhaustion, wedged controller managers that don't run reconcile loops but hold the lease, ...)
- Analysis of the severity of the possible failure modes. E.g. Is complete loss of controller reconciliation until a human intervenes possible in an HA cluster if a mistake is made in graceful shutdown?
- If the severity is high (my hunch is that there are high severity failure modes, but I'd like to see analysis), then what sort of technical safeguards can be put in place to prevent them?

Reply:
Expanded the section. Not really sure what kind of safeguards we can put to prevent them other than testing and catching the problems early.

Reply:
> Expanded the section. Not really sure what kind of safeguards we can put to prevent them other than testing and catching the problems early.

Detection of spurious threads stands out as powerful. If we detect extra gofuncs while running, self-terminating is both appropriate and helpful.